src/main/kotlin/me/owdding/lib/cosmetics/CosmeticManager.kt

```diff
@@ -0,0 +1,152 @@
+                    ).apply(it) { players, cosmetics ->
+                        val cosmetics = cosmetics.filter { (_, version) -> version >= COSMETIC_VERSION }
+                        this._cosmetics.putAll(cosmetics.associateBy(Cosmetic::id))
+                        this._players.putAll(players.map(CompletablePlayerData::complete).associateBy(PlayerData::uuid))
+                    }
+                },
+            )
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+        this._mlibCosmetics.putAll(
+            playerList.associateNotNull(
+                keySelector = { it.uuid },
+                valueSelector = {
+                    it.data.toData(MeowddingLibCodecs.getCodec<MlibCosmeticData>())?.takeUnless { data -> data == emptyCosmetic }
+                },
+            ),
+        )
+        MeowddingLib.info("Fetched ${players.size} players and ${cosmetics.size} cosmetics")
+    }
+
+    @GenerateCodec
+    internal data class CompletablePlayerData(
+        val uuid: UUID,
+        @FieldName("extra_data") val extraData: JsonObject,
+        val cosmetics: List<CosmeticId>,
+    ) {
+        fun complete() = PlayerData(
+            uuid,
+            JsonObject().apply {
+                extraData.entrySet().forEach { (key, value) -> add(key, value) }
+
+                cosmetics.mapNotNull { it.resolve() }.filter { it.version <= COSMETIC_VERSION }.forEach {
+                    it.data.entrySet().forEach { (key, value) ->
+                        if (key == "version" || key == "id") return@forEach
+                        if (has(key)) return@forEach
+                        add(key, value)
+                    }
+                }
+            },
+        )
+    }
+
+    data class PlayerData(
+        val uuid: UUID,
+        val data: JsonObject,
+    )
+
+    @GenerateCodec
+    data class Cosmetic(
+        val id: CosmeticId,
+        val version: Int,
+        @Unnamed val data: JsonObject,
+    )
+
+    @JvmInline
+    value class CosmeticId(val cosmeticId: String)
+
+    fun CosmeticId.resolve(): Cosmetic? = cosmetics[this]
+
+    @IncludedCodec
+    internal val COSMETIC_ID_CODEC: Codec<CosmeticId> = Codec.STRING.xmap(::CosmeticId, CosmeticId::cosmeticId)
+
+    @GenerateCodec
+    data class MlibCosmeticData(
+        @Lenient val suffix: Component?,
+        @NamedCodec("cosmetic_url_type") @Lenient @FieldName("cape_texture") val capeTexture: URI?,
+    )
+
+    private val emptyCosmetic = MlibCosmeticData(null, null)
+
+    @IncludedCodec(named = "cosmetic_url_type")
+    val URL_KEY: Codec<URI> = Codec.STRING.flatXmap(
+        { string ->
+            runCatching {
+                URI.create(string).takeUnless { it.host != "files.meowdd.ing" }
+            }.getOrNull()?.let {
+                DataResult.success(it)
+            } ?: DataResult.error { "Invalid cosmetic url!" }
+        },
+        { DataResult.success(it.toString()) },
+    )
```

Comment:
these classes and codecs should be put somewhere else, they should also be in a consistent order with variables separated from classes